Use `Word31` as underlying type for `EpochNo`.

jonathanknowles opened the pull request:
# Issue Number

#1025 

# Overview

This PR:

- [x] Changes the underlying type of `EpochNo` from `Word64` to `Word31`.
- [x] Adds validation to our Byron CBOR and Jörmungandr decoders that fail fast when encountering epoch numbers that are out of bounds.

# Outstanding work

- [x] Revise and/or simplify multiple `Arbitrary` instances for `SlotId` and `EpochNo`.
- [x] Verify that the `NFData` instance for `EpochNo` is correct.
- [x] Figure out why `fromFlatSlot . flatSlot` is failing, and fix the test (or the code).
- [x] Write tests to demonstrate that arbitrary values of `SlotId` can be persisted to the database and read back successfully.
- [ ] Write tests to demonstrate that binary decoders fail when given out-of-bounds epoch numbers.

jonathanknowles committed:
Use `Word31` as the underlying type for the `EpochNo` type.

- Add function `unsafeEpochNo`.
- Fail if we encounter an out-of-bounds epoch number while decoding.
- Add function `decodeSlotNo`: Use it to simplify function `decodeSlotId`.
- Adjust return type of `decodeSlotId` to `SlotId`.
- Use `unsafeEpochNo` to simplify conversions to `SlotNo` from other types.
- Give `EpochLength` a deconstructor.
- Include actual bounds and given value in error message of
  unsafeEpochNo: It increases a bit the user experience and debugging
  capability.
- Correct comment for `artificialEpochLength`.

Co-authored-by: KtorZ <matthias.benkort@iohk.io>

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -285,10 +287,10 @@ instance FromJSON SlotNo where
 
 instance FromJSON EpochNo where
-    parseJSON = fmap EpochNo . parseJSON
+    parseJSON = fmap (EpochNo . fromIntegral @Word32 @Word31) . parseJSON
```
We probably want a parse failure of some sort here, should the value be wider than 31 bits.

KtorZ replied:
Hmm No. These instances are not used and are "unsound". Sadly required by the persistent auto-generated machinery for a mysterious reason :disappointed:

KtorZ commented on lib/core/src/Cardano/Byron/Codec/Cbor.hs:
```diff
@@ -462,10 +477,10 @@ decodeSharesProof = do
     return ()
 
-decodeSlotId :: CBOR.Decoder s (Word64, Word32)
+decodeSlotId :: CBOR.Decoder s (EpochNo, Word32)
```
For consistency, I'd suggest to also make `Word32` a `SlotNo` here.

jonathanknowles replied:
> For consistency, I'd suggest to also make `Word32` a `SlotNo` here.

Done.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite/TH.hs:
```diff
@@ -55,7 +55,7 @@ PoolProduction sql=pool_production
 StakeDistribution sql=stake_distribution
     stakeDistributionPoolId     W.PoolId     sql=pool_id
-    stakeDistributionEpoch      Word64       sql=epoch
+    stakeDistributionEpoch      Word32       sql=epoch
```
Unfortunately, this now requires a database migration to not break every existing wallets out there. Keeping it as a `Word64` also works for now with a `TODO` tag.

jonathanknowles replied:
> Unfortunately, this now requires a database migration to not break every existing wallets out there. Keeping it as a `Word64` also works for now with a `TODO` tag.

Fair enough. I've reverted this to `Word64`.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -990,9 +993,15 @@ newtype SlotNo = SlotNo { unSlotNo :: Word32 }
+newtype EpochNo = EpochNo { unEpochNo :: Word31 }
     deriving stock (Show, Read, Eq, Ord, Generic)
-    deriving newtype (Num, Buildable, NFData, Enum)
+    deriving newtype (Num, Bounded, Enum)
```
So this is the source of your question about `NFData` :) 
Looking more closely at the source code, `OddWord` is actually a `newtype` constructor around a `Word32`. Which promises no particular runtime overhead of the type itself. Meaning that evaluating `EpochNo` to WHNF should be enough to fully evaluate it.

jonathanknowles replied:
@KtorZ wrote:

> So this is the source of your question about NFData :)

Yes!

>  Looking more closely at the source code, OddWord is actually a newtype constructor around a Word32. Which promises no particular runtime overhead of the type itself. Meaning that evaluating EpochNo to WHNF should be enough to fully evaluate it.

Fair enough, I think this argument makes sense.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1079,8 +1082,10 @@ instance Arbitrary SlotNo where
-    arbitrary = EpochNo <$> arbitrary
+    shrink (EpochNo x) =
+        EpochNo . fromIntegral @Word32 <$>
+            shrink (fromIntegral @Word31 x)
```
Alternatively, using [shrinkIntegral](https://hackage.haskell.org/package/QuickCheck-2.13.2/docs/Test-QuickCheck-Arbitrary.html#v:shrinkIntegral)

```hs
shrink (EpochNo x) = EpochNo <$> shrinkIntegral x
```

jonathanknowles replied:
Fixed.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:
```diff
@@ -308,8 +310,12 @@ instance Arbitrary SlotNo where
-    arbitrary = EpochNo <$> choose (0, fromIntegral arbitraryEpochLength)
+    shrink (EpochNo x) =
+        EpochNo . fromIntegral @Word32 <$>
+            shrink (fromIntegral @Word31 x)
```
Same as previous comment about [shrinkIntegral](https://hackage.haskell.org/package/QuickCheck-2.13.2/docs/Test-QuickCheck-Arbitrary.html#v:shrinkIntegral)

jonathanknowles replied:
Fixed.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -1073,7 +1073,7 @@ instance {-# OVERLAPS #-} Arbitrary (EpochLength, SlotId) where
     arbitrary = do
         (EpochLength epochLength) <- arbitrary
-        ep <- EpochNo <$> choose (0, 1000)
+        ep <- EpochNo . fromIntegral @Word32 <$> choose (0, 1000)
```
This transformation from `Word32` to `EpochNo` seems like a recurring enough operation that it could deserve its own function:

```hs
mkEpochNo :: Word32 -> EpochNo
```

This function is (ought to be) partial, which isn't the case of `fromIntegral` which could yield wrong result. It'd be better for both readability and safety to use such a smart constructor IMO, which would make sure to fail loudly when passed a `Word32` that is too big (very much like the CBOR decoder in another module).

jonathanknowles replied:
> This transformation from `Word32` to `EpochNo` seems like a recurring enough operation that it could deserve its own function:
> This function is (ought to be) partial.

Done. I've called the function `unsafeEpochNo`, to reflect the fact that it may fail if the specified value is too large.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -247,6 +252,17 @@ getBlockId = lookAhead getBlock *> label "getBlockId" (do
+        | otherwise =
+            EpochNo $ fromIntegral epochNo
+    maxEpochNo :: Word32
+    maxEpochNo = fromIntegral @Word31 $ unEpochNo maxBound
```
From previous comment, I'd suggest slightly transforming this into:

```hs

mkEpochNo :: HasCallStack => Word32 -> EpochNo
mkEpochNo epochNo 
    | epochNo > maxEpochNo =
        error "Epoch number out of bounds when decoding from binary."
    | otherwise =
        EpochNo $ fromIntegral epochNo
  where
    maxEpochNo :: Word32
    maxEpochNo = fromIntegral @Word31 maxBound
```

And use it consistently everywhere else.

jonathanknowles replied:
Done.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -999,9 +1005,27 @@ newtype SlotNo = SlotNo { unSlotNo :: Word32 }
+        EpochNo $ fromIntegral epochNo
+  where
+    maxEpochNo :: Word32
+    maxEpochNo = fromIntegral @Word31 $ unEpochNo maxBound
```
My first instinct was to locate this function within the `Unsafe` module. However, doing so introduces a circular import.

jonathanknowles committed:
Fix the "flatSlot . fromFlatSlot == id" property test.

And correct the names of roundtrip tests for `flatSlot` and `fromFlatSlot`.

Make `fromFlatSlot` fail if unable to decode a valid `SlotId`.

Co-authored-by: KtorZ <matthias.benkort@iohk.io>

jonathanknowles committed:
Add roundtrip database field persistence test for `SlotId`.

Co-authored-by: KtorZ <matthias.benkort@iohk.io>

KtorZ commented on lib/core/src/Cardano/Byron/Codec/Cbor.hs:
```diff
@@ -292,6 +296,13 @@ decodeDifficulty = do
 
+decodeEpochNo :: HasCallStack => CBOR.Decoder s EpochNo
+decodeEpochNo =
+    unsafeEpochNo . fromIntegral @Word64 @Word32 <$> CBOR.decodeWord64
```
:+1:

KtorZ commented on lib/core/src/Cardano/Byron/Codec/Cbor.hs:
```diff
@@ -462,12 +473,11 @@ decodeSharesProof = do
     return ()
 
-decodeSlotId :: CBOR.Decoder s (Word64, Word32)
+decodeSlotId :: CBOR.Decoder s SlotId
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -999,9 +1005,27 @@ newtype SlotNo = SlotNo { unSlotNo :: Word32 }
+unsafeEpochNo :: HasCallStack => Word32 -> EpochNo
+unsafeEpochNo epochNo
+    | epochNo > maxEpochNo =
+        error "unsafeEpochNo: epoch number out of bounds."
```
**[suggestion: user-experience]**
The error message could include the actual bounds and what was the given value.

**Why**
Increase a bit the user experience and debugging capability.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1138,15 +1162,32 @@ syncProgressRelativeToTime tolerance sp tip time =
+            , ") for the specified epoch length ("
+            , show epochLength
+            , ")."
+            ]
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/Sqlite/TypesSpec.hs:
```diff
@@ -0,0 +1,74 @@
+import Data.Word.Odd
+    ( Word31 )
+import Database.Persist.Sqlite
+    ( PersistField (..) )
```
**[suggestion: dependency management]**

Use [`PersistField`](https://hackage.haskell.org/package/persistent-2.10.4/docs/Database-Persist-Class.html#t:PersistField) from `persistent` instead

**Why**

This is the same function re-exported by `persistent-sqlite` however, using the version from `Database.Persist.Sqlite` creates a non-necessary dependency to `persistent-sqlite` whereas we are just interested in a class already available in the core of persistent.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -221,11 +232,33 @@ spec = do
+                if n <= maxFlatSlot then
+                    result `shouldBe` n
+                else
+                    evaluate result `shouldThrow` anyErrorCall
```
:+1:

KtorZ reviewed (approved):
LGTM. Some minor suggestions.

KtorZ committed:
add basic happy and negative tests for unsafeEpochNo to check its behavior

KtorZ committed:
re-generate nix machinery

KtorZ committed:
lower down generated values for UniformTimeRange

Previously, was going up to 1000 years which is ... overkill. Considering the worst-case scenario for
epoch, i.e., changing every second, it'll give us roughly 67 years before overflowing an epoch number
encoded on 31 bits. So, using 50 years as an upper bound here, just to make things more realistic
